### ISSUE TITLE
False negative: Protocols not checked in return types

### ISSUE THREAD

**Issue by patrick-kidger** [python/mypy#12843]
**To Reproduce**

The following is not rejected by mypy:

```python
from typing import Protocol, TypeVar

T = TypeVar("T")
S = TypeVar("S")

class FooProtocol(Protocol[T, S]):
    bar: T
    baz: S

class Foo:
    def __init__(self, bar, baz):
        self.bar = bar
        self.baz = baz

def swap(x: FooProtocol[T, S]) -> FooProtocol[S, T]:
    return Foo(x.baz, x.baz)  # Typo! But no error raised.

swap(Foo(5, "hi"))
```

**Expected Behavior**

mypy should reject the above: the actual return "type" (protocol) of `swap` is `FooProtocol[S, S]`, rather than the declared `FooProtocol[S, T]`.

**Your Environment**

- Mypy version used: 0.95.0
- Python version used: 3.10.0
- Operating system and version: Ubuntu 20.04 on WSL2 on Windows 10

**Additional Information**

The corresponding nominal subtyping approach is correctly rejected by mypy:

```python
from typing import Generic, TypeVar

T = TypeVar("T")
S = TypeVar("S")

class Foo(Generic[T, S]):
    def __init__(self, bar: T, baz: S):
        self.bar = bar
        self.baz = baz

def swap(x: Foo[T, S]) -> Foo[S, T]:
    return Foo(x.baz, x.baz)

swap(Foo(5, "hi"))

# tmp5.py:12: error: Argument 2 to "Foo" has incompatible type "S"; expected "T"
```

For context: I'm the author of [TorchTyping](https://github.com/patrick-kidger/torchtyping), so I've been playing around with mypy in anticipation of 3.11 and variadic generics making mypy compatibility a future possibility. Unfortunately I found that a misimplemented "transpose array" function (similar to the above `swap`) did not raise the expected error.

**Comment by patrick-kidger** [comment]
I have noticed that defining
```python
class Foo(Generic[T, S]):
    def __init__(self, bar: T, baz:S ):
        self.bar = bar
        self.baz = baz
```
correctly raises an error, which is good.

But to my mind (or at least my use cases) protocols -- more generally the use of structural subtyping of nominal subtyping -- are mostly useful when another library has authored `Foo`, and as such you can't change it, so the above fix doesn't apply.

**Comment by JelleZijlstra** [comment]
This is expected behavior. `Foo.__init__` is untyped, so we don't check its arguments.

**Comment by patrick-kidger** [comment]
Could you describe the appropriate way to tackle this issue then, please? Given that `Foo` is defined in some external library and cannot be modified.

**Comment by JelleZijlstra** [comment]
Provide stub files for the library that Foo is in.

**Comment by erictraut** [comment]
FWIW, pyright reports a type violation in this circumstance. But if you want to stick with mypy, stubs are the way to go here.
